### PR TITLE
First draft of add deadline to todo

### PR DESCRIPTION
**Pull request by theDisco**
The front end is very primitive, but it is a first shot at understanding CQRS. After merging with upstream it also seems that something is wrong with doctrine.

### CODE REVIEW

File: src/Model/Todo/Event/DeadlineWasAddedToTodo.php [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+     */
+    private $deadline;
+
+    /**
+     * @param TodoId $todoId
+     * @param UserId $userId
+     * @param \DateTime $deadline
+     * @return DeadlineWasAddedToTodo
+     */
+    public static function byUserToDate(TodoId $todoId, UserId $userId, \DateTime $deadline)
```
Comment (prolic) [inline src/Model/Todo/Event/DeadlineWasAddedToTodo.php]: please use datetimeimmutable

Comment (theDisco) [inline src/Model/Todo/Event/DeadlineWasAddedToTodo.php]: sure

File: src/Model/Todo/Command/AddDeadlineToTodo.php [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+class AddDeadlineToTodo extends Messaging\Command implements Messaging\PayloadConstructable
+{
+    use Messaging\PayloadTrait;
+
+    /**
+     * @return string
+     */
+    public function userId()
+    {
+        return $this->payload['user_id'];
```
Comment (prolic) [inline src/Model/Todo/Command/AddDeadlineToTodo.php]: return object userId here

File: src/Model/Todo/Command/AddDeadlineToTodo.php [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+    }
+
+    /**
+     * @return string
+     */
+    public function todoId()
+    {
+        return $this->payload['todo_id'];
```
Comment (prolic) [inline src/Model/Todo/Command/AddDeadlineToTodo.php]: return object TodoId

File: src/Model/Todo/Command/AddDeadlineToTodo.php [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+    }
+
+    /**
+     * @return string
+     */
+    public function deadline()
+    {
+        return $this->payload['deadline'];
```
Comment (prolic) [inline src/Model/Todo/Command/AddDeadlineToTodo.php]: return object again

File: src/Model/Todo/Event/DeadlineWasAddedToTodo.php [hunk 5]
```diff
@@ -0,0 +1,87 @@
...
+use Prooph\EventSourcing\AggregateChanged;
+use Prooph\ProophessorDo\Model\Todo\TodoId;
+use Prooph\ProophessorDo\Model\User\UserId;
+
+/**
+ * Class DeadlineWasAddedToTodo
+ * @package Prooph\ProophessorDo\Model\Todo\Event
+ * @author Wojtek Gancarczyk <wojtek@aferalabs.com>
+ */
+class DeadlineWasAddedToTodo extends AggregateChanged
```
Comment (prolic) [inline src/Model/Todo/Event/DeadlineWasAddedToTodo.php]: make it final

File: src/Model/Todo/Handler/AddDeadlineToTodoHandler.php [hunk 6]
```diff
@@ -0,0 +1,45 @@
...
+use Prooph\ProophessorDo\Model\Todo\TodoId;
+use Prooph\ProophessorDo\Model\Todo\TodoList;
+
+/**
+ * Class AddDeadlineToTodoHandler
+ *
+ * @package Prooph\ProophessorDo\Model\Todo\Handler
+ * @author Wojtek Gancarczyk <wojtek@aferalabs.com>
+ */
+class AddDeadlineToTodoHandler
```
Comment (prolic) [inline src/Model/Todo/Handler/AddDeadlineToTodoHandler.php]: make it final

**Comment by prolic** [comment]
looks good in first sight, though i din't test it

**Comment by theDisco** [comment]
I have moved the factories and type initialization to the command.

**Comment by prolic** [comment]
php-cs issues again ;)

**Comment by theDisco** [comment]
Done, added the final keyword to classes as requested.

**Comment by prolic** [comment]
Will check again this evening.

File: src/Model/Todo/Handler/AddDeadlineToTodoHandler.php [hunk 7]
```diff
@@ -0,0 +1,44 @@
...
+    /**
+     * @param AddDeadlineToTodo $command
+     * @return void
+     * @throws \Exception
+     */
+    public function __invoke(AddDeadlineToTodo $command)
+    {
+        $todo = $this->todoList->get($command->todoId());
+
+        if ($todo->assigneeId()->toString() !== $command->userId()->toString()) {
```
Comment (codeliner) [inline src/Model/Todo/Handler/AddDeadlineToTodoHandler.php]: Move this check into `Todo::addDeadline` method by passing the userId from the command to the Todo. 
Background: The aggregate is responsible for protecting invariants. The handler only locates the correct aggregate and triggers the action.

Also a dedicated exception class would be nice. See [existing exceptions](https://github.com/prooph/proophessor-do/tree/master/src/Model/Todo/Exception) for inspiration.

File: src/Model/Todo/Todo.php [hunk 8]
```diff
@@ -71,6 +77,24 @@ public function markAsDone()
     }
 
     /**
+     * @param \DateTimeImmutable $deadline
+     * @return void
+     * @throws \Exception
+     */
+    public function addDeadline(\DateTimeImmutable $deadline)
+    {
+        $this->recordThat(DeadlineWasAddedToTodo::byUserToDate($this->todoId, $this->assigneeId, $deadline));
```
Comment (codeliner) [inline src/Model/Todo/Todo.php]: Before you record the event you should make sure that the deadline is valid: Should not be a date in the past!

See also comment above and move the user assertion to this location.

**Comment by codeliner** [comment]
@theDisco great job so fare :+1: I've commented some things. But should be easy to correct them.
Regarding the UI: Your implementation looks fine. I may add a jquery calendar, but you can focus on the backend work.

**Comment by theDisco** [comment]
@codeliner I have made some changes and I need your input. I basically moved all the checks, as you've suggested, to the Todo model. I wasn't entirely sure about adding the `created_on` to the payload. Correct me if I'm wrong, but when upcastig, if the TodoDeadling would compare the deadline date with the current date, it could come to a case, where the check will fail, although the payload was completely valid in moment of execution. Therefor I need `created_on` date, to be able to correctly validate the deadline.

**Comment by codeliner** [comment]
@theDisco The changes look great :+1: especially that you've added a TodoDeadline VO.

Regarding your question: 

"upcasting" wouldn't be a problem at all. "upcasting events" means the event stream is modified in case that a new application version requires changes in the structure of events.

What you are looking for is called "replay" events. But also in this case it is no problem if the TodoDeadline compares the deadline date with the current date. It is only important that `TodoDeadline::isInThePast` performs a correct check at the moment it used by the `Todo` to validate the deadline. Also when the Todo will use the `TodoDeadline` VO later to check if the deadline is expired it cannot use the `TodoDeadline` for this check at the moment as the `isInThePast` method will always return false because it compares the deadline always against the internal `created_on` datetime.

In an event sourced aggregate it is very important that validation takes place before domain events are recorded. When the same events are replayed validation logic is not executed again. This is the reason why these `when<event name>` methods exist. In your case for example the [whenDeadlineWasAddedToTodo](https://github.com/theDisco/proophessor-do/blob/dc9e3dd1a9989ec1c83cfe28aef0ee6e543a9df5/src/Model/Todo/Todo.php#L158) method.
Only this method is called when the `DeadlineWasAddedToTodo` event is replayed.

So I would say you can or even should remove the `created_on` datetime from the payload but keep in mind that the `TodoDeadline::isInThePast` check needs to consider the client timezone to compare the current datetime with deadline datetime.

**Comment by theDisco** [comment]
> So I would say you can or even should remove the created_on datetime from the payload but keep in mind that the TodoDeadline::isInThePast check needs to consider the client timezone to compare the current datetime with deadline datetime.

I have removed the created_on from the payload but still keep it internally to provide verbose error messages in the exception. Regarding the TZ. Should it come directly from server settings (aka date_default_timezone_get) or should it be something set on a user level and obtained directly from user model?

**Comment by codeliner** [comment]
@theDisco :+1: I will check your changes in the evening. Regarding TZ. This should be taken from the user request because the server does not know which timezone is configured in the client. The client would need to send a `ISO8601` formated deadline date. On server side you could then derive the timezone from the deadline date by using: http://php.net/manual/de/datetime.gettimezone.php and pass this timezone to a new datetime instance: 
```php
$now = new \DateTimeImmutable('now', $deadline->getTimezone());
```

Would be ok if you only implement the server part. I will add a date picker to the UI anyway and can then also send the ISO date.

**Comment by codeliner** [comment]
@theDisco I'm going to merge. Did a final review and your PR looks really great :+1: 
I'll check the timezone handling when adding the date picker. No need to add it with this PR.
Thank you again for tackling the task.

**Comment by theDisco** [comment]
@codeliner Thanks! I was going to implement the TZ handling but I was very busy last couple days. I will try to do it on weekend.

**Comment by codeliner** [comment]
@theDisco ok cool. I prepare the UI part so that you only need to handle backend side.

**Pull request merged**